### ISSUE TITLE
Dynamic ft_transfer_calls in FT standard implementation

### ISSUE THREAD

**Issue by mfornet** [near/near-sdk-rs#619]
Current implementation of `ft_transfer_call` <span class="error">[allocates exactly 25TGas]</span>([https://github.com/near/near-sdk-rs/blob/master/near-contract-standards/src/fungible_token/core_impl.rs#L12](https://github.com/near/near-sdk-rs/blob/master/near-contract-standards/src/fungible_token/core_impl.rs#L12)) for the tx itself. In the case that less gas is required, the reminder gas can't be used for the rest of the transfer. But most importantly is the case when this gas is not enough to run the transfer, which can happen when the payload attached is too large. See [https://github.com/aurora-is-near/eth-connector/issues/27](https://github.com/aurora-is-near/eth-connector/issues/27).

I propose to determine the gas used dynamically, in such a way that the gas attached to subsequence cross contract calls is determined from the gas remaining at the end of the function.

**Comment by itegulov** [comment]
@austinabell although NEP264 will help with the first half of the issue (not being able to use the leftover gas for the rest of the transfer), I don't think it will help with the second one (running out of gas before calling `ft_on_transfer` despite passing the initial gas check which is arguably more severe). We would have to make `ft_transfer_call` estimate how much gas it takes dynamically from the payload size and in general a part of a bigger problem with static gas estimation tactics.

**Comment by austinabell** [comment]
> @austinabell although NEP264 will help with the first half of the issue (not being able to use the leftover gas for the rest of the transfer), I don't think it will help with the second one (running out of gas before calling `ft_on_transfer` despite passing the initial gas check which is arguably more severe). We would have to make `ft_transfer_call` estimate how much gas it takes dynamically from the payload size and in general a part of a bigger problem with static gas estimation tactics.

This is true. Perhaps the solution we need is to expose is a way to have more granular control over gas used without breaking the simple path?

Although yes it doesn't solve the issue entirely, I don't see the issue with failing before calling `ft_on_transfer`. Yes, the transaction will fail, but none of the async calls will go off so it's much more acceptable. `ft_transfer_call` estimating how much gas it takes dynamically doesn't seem that feasible since it relies on current gas prices for operations and payload size is only one factor.

Also, we can't schedule the callback using the remaining gas after each receipt execution as suggested by the issue because it requires the contract implementing `ft_on_transfer` to schedule the callback, which they can just omit if they wanted.

We can use NEP264 to make this much less of an issue by keeping the static gas (and possibly adding a small weight of unused gas) for the callback and just allocating most unused gas to the `ft_on_transfer` while allowing a developer to modify each. If the `ft_on_transfer` call is important, they can add a static amount of gas or increase the ratio, same with `ft_resolve_transfer`.

**Comment by itegulov** [comment]
All great points, thanks for the elaborate response.

> while allowing a developer to modify each. If the ft_on_transfer call is important, they can add a static amount of gas or increase the ratio, same with ft_resolve_transfer.

The only thing that is unclear to me is in what form the developer would be allowed to modify these values. Wouldn't they be hardcoded in some form in contract standards? I don't really see how one would be able to change static amount of gas or ration without copying the code and changing the values there.
